docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,279 @@
+  // errors that can be converted to api.Status. It returns nil if the object should be created.
+  BeforeCreate(ctx api.Context, obj runtime.Object) error
+
+  // Validate ensures that a resource is valid on creation. It is invoked after BeforeCreate.
+  Validate(ctx api.Context, obj runtime.Object) error
+
+  // Create creates a new version of a resource.
+  Create(ctx api.Context, obj runtime.Object) (runtime.Object, error)
+}
+```
+
+### Current
+
+[apiserver create flow](https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/apiserver/resthandler.go#L333)
+1. decode incoming request
+1. admit.Admit
+1. r.Create (the items below are part of the generic registry store / strategy-based storage)
+  1. rest.BeforeCreate
+    1. strategy.PrepareForCreate
+    1. api.FillObjectMetaSystemFields
+    1. api.GenerateName
+    1. strategy.Validate
+    1. validation.ValidateObjectMeta
+    1. strategy.Canonicalize
+  1. e.Storage.Create
+  1. e.AfterCreate
+  1. e.Decorator
+
+### Proposed apiserver create flow
+
+1. decode incoming request
+1. admit.ApplyDefaults
+1. admit.Authorize
+1. r.BeforeCreate
+  1. If strategy-based, call rest.BeforeCreate
+    1. strategy.PrepareForCreate
+    1. api.FillInObjectMetaSystemFields
+    1. api.GenerateName
+  1. Otherwise, just do whatever is in the rest.Creater
+1. r.Validate
+  1. If strategy-based, call (new) rest.Validate
+    1. strategy.Validate
+    1. validation.ValidateObjectMeta
+  1. Otherwise, just do whatever is in the rest.Creater
+1. admit.Validate
+  1. e.g. for quota check
+1. r.Create
+  1. For strategy-based handlers:
+    1. strategy.Canonicalize
+    1. e.Storage.Create
+    1. e.AfterCreate
+    1. e.Decorator
+  1. Otherwise, just do whatever is in the rest.Creater
+
+### Proposed apiserver update flow (TODO finish this section)
+
+1. decode request
+1. admit.ApplyDefaults
+1. admit.Authorize
+1. r.BeforeUpdate
+1. r.ValidateUpdate
+1. admit.PostValidation
+1. r.Update
+
+### Proposed apiserver patch flow (TODO finish this section)
+
+1. decode request
+1. admit.
+
+### Proposed apiserver delete flow (TODO finish this section)
+
+1. decode request (DeleteOptions)
+1. admit.Authorize?
+1. admit.PostValidation
+1. r.Delete
+
+### Proposed apiserver delete collection flow (TODO finish this section)
+
+1.
+
+### Proposed apiserver connect flow (TODO finish this section)
+
+1. admit.Authorize
+1. connector.Connect
+
+
+## Ability to run out-of-tree plugins
+
+Admission plugins are all currently compiled into the apiserver binary. If you want to run custom plugins, you have to modify the Kubernetes source code and recompile the apiserver. Some examples of custom plugins that an administrator might want to run (these are all actual plugins in OpenShift today):
+
+- default restart-never and restart-on-failure pods' `activeDeadlineSeconds` based on either a namespace-level annotation, or the plugin's configured value.
```

Comment:
@bparees the build-related admission plugins only affect OpenShift resources, right? In other words, if we end up running a standalone OpenShift apiserver, we would just compile build admission into it, and it wouldn't need to be an out-of-tree plugin that the kube-apiserver invoked. Correct?

@smarterclayton @liggitt @deads2k @derekwaynecarr @pmorie do we have any other examples of admission plugins in OpenShift that have to run as part of the kube-apiserver request lifecycle?